packages/google_sign_in/lib/src/oauth2.dart

```diff
@@ -0,0 +1,276 @@
+class DeviceAuthClient {
+  /// Creates an instance of [DeviceAuthClient].
+  DeviceAuthClient({
+    required this.authorizationEndPoint,
+    required this.tokenEndPoint,
+    required this.revokeEndPoint,
+    http.Client? httpClient,
+  }) : _httpClient = httpClient ?? http.Client();
+
+  /// The server endpoint that returns authroization grant when user grants access.
+  final Uri authorizationEndPoint;
+
+  /// The server endpoint that returns access token when given authorization grant.
+  final Uri tokenEndPoint;
+
+  /// The server endpoint that revokes future access to issued tokens.
+  final Uri revokeEndPoint;
+
+  final http.Client _httpClient;
+
+  bool _isPolling = false;
+
+  /// Checks whether polling started by by [pollToken] is in progress.
+  bool get isPolling => _isPolling;
+
+  /// Stops poll request started by [pollToken].
+  void cancelPollToken() => _isPolling = false;
+
+  /// Requests authroization grant from [authorizationEndPoint].
+  Future<AuthorizationResponse> requestAuthorization(
+      String clientId, List<String> scope) async {
+    final Map<String, String> body = <String, String>{
+      'client_id': clientId,
+      'scope': scope.join(' '),
+    };
+
+    final http.Response response =
+        await _httpClient.post(authorizationEndPoint, body: body);
+
+    final Map<String, dynamic> jsonResponse =
+        convert.jsonDecode(response.body) as Map<String, dynamic>;
+    if (response.statusCode != 200) {
+      _handleErrorResponse(jsonResponse);
+    }
+
+    return AuthorizationResponse.fromJson(jsonResponse);
+  }
+
+  /// Requests tokens from [tokenEndPoint].
+  Future<TokenResponse> requestToken(
+    String clientId,
+    String clientSecret,
+    String deviceCode,
+  ) async {
+    final Map<String, String> body = <String, String>{
+      'grant_type': 'http://oauth.net/grant_type/device/1.0',
+      'client_id': clientId,
+      'client_secret': clientSecret,
+      'code': deviceCode,
+    };
+
+    final http.Response response =
+        await _httpClient.post(tokenEndPoint, body: body);
+    final Map<String, dynamic> jsonResponse =
+        convert.jsonDecode(response.body) as Map<String, dynamic>;
+
+    if (response.statusCode != 200) {
+      _handleErrorResponse(jsonResponse);
+    }
+
+    return TokenResponse.fromJson(jsonResponse);
+  }
+
+  /// Repeat sending token request to [tokenEndPoint] until user grants access.
+  Future<TokenResponse?> pollToken({
+    required String clientId,
+    required String clientSecret,
+    required String deviceCode,
+    required Duration interval,
+  }) async {
+    if (isPolling) {
+      throw PlatformException(
+        code: 'already-polling',
+        message: 'Client is already polling token from server, cancel the '
+            'previous poll request before starting a new one.',
+      );
+    }
+    _isPolling = true;
+    Duration currentInterval = Duration(seconds: interval.inSeconds);
```

Comment:
`interval` is not final so you shouldn't need a copy of it.

If you think you need, you can just write

```dart
Duration currentInterval = interval;
```

Reply:
Right, fixed!